packages/block-library/src/search/index.php

```diff
@@ -119,6 +135,39 @@ function register_block_core_search() {
 }
 add_action( 'init', 'register_block_core_search' );
 
+/**
+ * Build an array with inline styles defining the border radius
+ * which will be applied to the inner elements and wrapper of the search block
+ * on the frontend.
+ *
+ * @param  array $attributes Search block attributes.
+ * @return array Border radius inline styles.
+ */
+function block_core_search_build_css_border_radius( $attributes ) {
+	$padding       = 4;
+	$border_radius = array(
+		'style'         => '',
+		'wrapper_style' => '',
+	);
+
+	$has_border_radius = isset( $attributes['style']['border']['radius'] );
+	if ( ! $has_border_radius ) {
+		return $border_radius;
+	}
+
+	$radius_value  = $attributes['style']['border']['radius'];
+	$button_inside = array_key_exists( 'buttonPosition', $attributes )
+		&& 'button-inside' === $attributes['buttonPosition'];
+
+	// Uses min value of 1 as having no radius when the outer does is jarring.
+	$radius = $button_inside ? max( $radius_value - $padding, 1 ) : $radius_value;
+
+	$border_radius['style']         = sprintf( 'border-radius: %spx;', esc_attr( $radius ) );
+	$border_radius['wrapper_style'] = sprintf( 'border-radius: %spx;', esc_attr( $radius_value ) );
+
+	return $border_radius;
+}
```

Comment:
I was hoping we'd avoid the need for any particular code on blocks when you enable "support" for this? why is this needed?

Reply:
This was required to maintain visual consistency between the border radius for the block itself and its inner elements when appropriate. 

When the search block is configured to position the button "inside", the border radius that the search block opted into needs to be applied not only to the wrapper around the search input and button but also to the input and button inner elements. This is complicated a little further as they need to be adjusted for the inner padding of the wrapper so that the border radius can match visually with the wrapper's border radius and we do not get "fat" corners.

This addresses an [issue highlighted](https://github.com/WordPress/gutenberg/pull/25203#issuecomment-689870244) with the original PR that sparked this one.

Reply:
Do you think it's possible to leverage some CSS magic to fix that instead? I'm asking cause this makes me wonder whether the block support is the right approach for this particular block or not. Basically, why is it even a block support if it requires such custom code, it removes the declarativeness of the block support. 

Block supports are and should be deterministic IMO: you apply them, you get a block radius automatically applied without custom tweaks. So it makes wonder whether block radius support flag is better suited for simpler blocks say button for instance or a group...

Reply:
> Do you think it's possible to leverage some CSS magic to fix that instead?

That's a good question, I'll look into that. Not being able to use an inherited value within `calc()` might make it tricky. Maybe CSS variables could offer something. 

One other thing that could complicate this, arose from some internal discussions, which raised the possibility of adding spacing/padding and border width support, both of which would affect the calculations for the inner border radii.

> I'm asking cause this makes me wonder whether the block support is the right approach for this particular block or not. Basically, why is it even a block support if it requires such custom code, it removes the declarativeness of the block support.

That's true regarding the search block itself. However, as you mention, simpler blocks could benefit from the border radius support and keep that declarative nature. If the search block still needs to offer a border radius feature, would it not be better to reduce duplication of code and leverage the block support for border radius?

We still have the initial PRs trialling adding the border radius feature for the search block in an ad-hoc manner, https://github.com/WordPress/gutenberg/pull/25553 & https://github.com/WordPress/gutenberg/pull/25203.

If you think that approach is actually the better option for the search block, I'm happy to remove the modifications to that block from this PR. It would be a shame though not to proceed with adding border radius support that other blocks can start to use.

Reply:
I'd love some thoughts from friends @nosolosw @jorgefilipecosta @mcsf @mtias

Reply:
I’ve tried utilising a CSS approach to achieve the desired results in the search block however it leaves a lot to be desired. 

**Approach:**
* Each block can have it’s own border radius. This means each block needs it’s own CSS rule to set a  `--border-radius` CSS variable within a selector that targets that block only.
* With a properly scoped CSS variable, the input and button inner components could have generic CSS rules that would inherit that value using something like: 
`border-radius: max( calc( var( --border-radius ) - 4px ), 1 );`
* The downside is that each block would now need to generate its own `<style>`  and CSS rule to achieve the above. Still moving away from the ideal where use of the border radius support for the search block was only declarative.

**Alternate approaches:**
* If we used preset values in the theme.json, global styles would handle injection of those CSS variables into the stylesheet. We then lose the ability to choose custom values and adding 50 preset values isn’t an option. 
* To use the border radius block support directly on the inner components, they would need to be converted to blocks that could opt in to the support. Operating independently, it would be up to the user to set the border radii on them, making sure they were in sync and manually calculating or visually adjusting the inner components’ radii until they achieved visual consistency with the outer wrapper radius when the button is placed inside. Not a great experience for the user.

**Summary:**

Presently, I don’t see an option to be able to provide the border radius feature on the search block as needed, without changes to the block itself. The current approach in this PR keeps those changes relatively straightforward, to a minimum and avoids duplicate code adding border radius controls etc that would increase the maintenance burden.

Happy to hear others’ opinions and any ideas to address concerns. For the moment though my vote would be to proceed with the current approach.

Reply:
This is a very tricky case.


When we were discussing if we should apply styles directly with global styles/ block support styles or applying variables `--wp-border-radius`, that the blocks would be responsible for dealing with, for this case it seems it would be easier if we followed the variables path. But if we used variables we would have the inheritance problem e.g: applying a border-radius for a group block would be inherited by the search block inside.


Under normal circumstances when we apply a border-radius there is no expectation that the child nodes will have a border-radius and that this border-radius is calculated based on the parent value. So what we have here is not a normal border-radius.


The block.json specifies support for border-radius, when global styles adds border-radius, it will be possible to control border-radius for the search block from global styles.
The solution works very well when we apply things at the local level, but it will not work when we apply a border via global styles. The child nodes would not have the border-radius style. We would have a border-radius property applied via CSS but the block border-radius style the server sees during the block rendering would be empty. We would have an inconsistency a search block with 30px border-radius applied at block level would look different than search block with 30px applied specifically to search block at the global styles level.


I think a block should never use a style attribute as part of its rendering, ideally, everything that comes from the style attribute can also come globally from global styles. If the block accesses the style attribute for something it creates an inconsistency between what happens with local styles and what happens with global styles.


I see three possible paths forward:

- We say a generic border-radius mechanism does not fit the needs of the search block. The search block does not uses any supports for border-radius and implements a border-radius picker at the block level, saved outside of the style attribute. I guess we may reuse components from a generic border-radius picker and the search block implementation to avoid duplicate code. The search block does not support global styles border-radius. A global styles setting will not be available for the block. If themes add it via theme.json it would also not work. On the theme template, HTML file themes can add a local border-radius if they use a search block there.


- We say the search block does not support border-radius but supports a style specific to it that is applied using a CSS variable. Similar to the way link color works. We would have something like: `--wp--style--block--core--search--border-radius` or `--wp--style--border--search-border-radius` that contains the border-radius specifically for the search block. The value is saved as part of the style attribute and the attribute is simply applied via a CSS variable. The block CSS uses the variable value to do its magic apply it to child nodes and do some calculations.
Themes would not be able to use the normal border-radius on theme.json to control the search block radius but they would be able to control the border radius of the search block using something custom like `search: { borderRadius: "50px" }` or `{searchBorderRadius: "50px" }`. If we follow this path I guess the first version could have some things hardcoded as the link color does. But in the future, we would have an API that allows third parties to add this type of specific styles. Essentially blocks would be able to add global styles that make sense to them.

- Blocks can say a specific style that already exists should be applied to them using a CSS variable instead of the default mechanism. When that flag is present both global styles and the local styles instead of generating "border-radius: 50px" on the presence of a border-radius in the search block scope would automatically generate something like `--wp--block--style--core--search--border-radius`. This variable is block specific so there is no inheritance. This is a generic system any block can opt-in to use variables for specific style. Themes would be able to apply border-radius style to search via theme.json as they apply to any other block (provided the style is scoped to the search block).

I am not sure what is the best path forward as none of the solutions seem perfect. The safest bet is the first option, it does not expose anything as part of our API's, but it is also the least powerful and discards totally the possibility of controlling border-radius from the search block from global styles/theme.json.
Any thoughts on this @mtias, @shaunandrews, @aristath, @nosolosw, @youknowriad?

Reply:
I honestly don't see any benefit to adding a border-radius control to the search block...
Border-radius is something that should be consistent on all elements on a page. Having different border-radii in different elements on a page creates a visual discontinuity and is a big no-no design-wise. It's OK to have some elements rounded while others are square(ish), but having a border-radius of 7px on some elements while other elements on the same page have 3px is not something that should be done. 
With that in mind, I'd say that themes should add a global border-radius in their theme.json file and use it consistently across the board for all their blocks - directly in their CSS files. The search block can have a separate "rounded" style, but there's no need for a border-radius control... It can simply be a block-style.

Reply:
I agree with the need for consistency of border radius across different elements on the page. 

One point worth noting here though is that differing border-radius can actually be needed to maintain visual consistency. For example, so that nested border radii actually appear consistent with each other rather than having "fat corners". 

An image highlighting this can be found in [this comment](https://github.com/WordPress/gutenberg/pull/25203#issuecomment-689870244).

Reply:
Yeah, but that is for elements in the same block.... So it's something that can be done directly in the theme's stylesheet, and values can be calculated easily via CSS calc. However, I was referring to cross-block inconsistencies. We can't have galleries use a 10px border-radius, search using a 6px radius, 1 group using 20px and search using something different... There needs to be consistency. And that consistency can be accomplished with simple CSS styles on the theme itself - without the need for any extra controls.

Adding a control to change the border-radius of a specific element will ultimately lead to visual inconsistencies since it won't follow the overall tone set by the theme. The search-block should either follow the global theme's border-radius (taking into account nested border-radii where needed in the theme's styles), or have a rounded style. Anything other than that would cause visual inconsistencies. So I really don't see why this can't be a simple default/rounded block-style...

Reply:
>  However, I was referring to cross-block inconsistencies. 

Sorry, that is what I meant by I agree with the need for consistency across different elements on the page. I was only trying to highlight that not all border radius values would be the same everywhere.

> Adding a control to change the border-radius of a specific element will ultimately lead to visual inconsistencies since it won't follow the overall tone set by the theme.

It could well do, but to play devil's advocate, should the end user have the freedom? We allow the option for users to choose random colors potentially destroying the look and feel of a theme.

> The search-block should either follow the global theme's border-radius (taking into account nested border-radii where needed in the theme's styles), or have a rounded style. Anything other than that would cause visual inconsistencies. So I really don't see why this can't be a simple default/rounded block-style...

This arose while attempting to build new block patterns that included border radii within the design for the pattern. I'll have to confirm the pattern previews can actually inherit from the global styles. If we are essentially tying the end user's hands with regards to the border radius values once the pattern is inserted, we would need to limit any disjoint between what they see in the preview vs the end result. That should be the case anyway but would be of lesser consequence if they have full control.

Reply:
> but to play devil's advocate, should the end user have the freedom?

We're about to enter a more philosophical discussion if we go down that path, but there is such a thing as too much control and a difference between control and enabling poor design choices. 

> This arose while attempting to build new block patterns that included border radii within the design for the pattern.

Block-patterns _can_ use block-styles. And a block-style simply adds a CSS-class to the block. So having a `is-style-round` class added to the search block is all that's really needed - which can be accomplished with a plain block-style selector like buttons have, and simple CSS :wink: